hat/app/org/hatdex/hat/api/controllers/devices/MachineData.scala

```diff
@@ -0,0 +1,338 @@
+    dataService: RichDataService,
+    loggingProvider: LoggingProvider,
+    configuration: Configuration,
+    usersService: UsersService,
+    trustedApplicationProvider: TrustedApplicationProvider,
+    implicit val ec: ExecutionContext,
+    implicit val applicationsService: ApplicationsService
+  )(wsClient: WSClient)
+    extends HatApiController(components, silhouette) {
+  import RichDataJsonFormats._
+
+  private val logger             = loggingProvider.logger(this.getClass)
+  private val defaultRecordLimit = 1000
+
+  val deviceVerification = new DeviceVerification(wsClient, configuration)
+
+  case class DeviceDataCreateRequest(body: Option[JsValue])
+  implicit val deviceDataCreateRequestReads: Reads[DeviceDataCreateRequest] = Json.reads[DeviceDataCreateRequest]
+
+  case class DeviceDataUpdateRequest(body: Seq[EndpointData])
+  implicit val machineDataUpdateRequestReads: Reads[DeviceDataUpdateRequest] = Json.reads[DeviceDataUpdateRequest]
+
+  // -- Error Handler - Error to Response
+  def handleFailedRequestAssessment(failure: DeviceVerificationFailure): Future[Result] =
+    failure match {
+      case DeviceRequestFailure(errorMessage) =>
+        logger.info(s"DeviceRequestFailure: ${errorMessage}")
+        Future.successful(BadRequest(s"DeviceRequest was invalid: ${errorMessage}"))
+
+      case ServiceRespondedWithFailure(errorMessage) =>
+        logger.info(s"ServiceRespondedWithFailure: ${errorMessage}")
+        Future.successful(ServiceUnavailable)
+
+      case ApplicationAndNamespaceNotValid =>
+        logger.info(s"ApplicationAndNamespaceNotValid")
+        Future.successful(BadRequest(s"DeviceRequest was invalid"))
+
+      case FailedDeviceRefinement =>
+        logger.info(s"FailedDeviceRefinement")
+        Future.successful(BadRequest(s"DeviceRequest was invalid."))
+
+      case InvalidTokenFailure(errorMessage) =>
+        logger.info(s"InvalidTokenFailure: ${errorMessage}")
+        Future.successful(BadRequest(s"DeviceRequest was invalid: ${errorMessage}"))
+
+      case InvalidDeviceDataRequestFailure(errorMessage) =>
+        logger.info(s"InvalidDeviceDataRequestFailure: ${errorMessage}")
+        Future.successful(BadRequest(s"DeviceRequest Body was invalid"))
+    }
+
+  // --- API Handlers ---
+
+  def getData(
+      namespace: String,
+      endpoint: String,
+      orderBy: Option[String],
+      ordering: Option[String],
+      skip: Option[Int],
+      take: Option[Int]): Action[AnyContent] =
+    UserAwareAction.async { implicit request =>
+      val eventuallyVerdict =
+        deviceVerification.getRequestVerdict(request.headers,
+                                             request.dynamicEnvironment.hatName,
+                                             namespace,
+                                             usersService,
+                                             trustedApplicationProvider
+        )
+
+      eventuallyVerdict.flatMap { verdict =>
+        verdict match {
```

Comment:
I see that you repeatedly use this unnecessary verbose syntax with a match on the lambda's parameter
it should be: 
```
eventuallyVerdict.flatMap {
    case Left(...) => ...
}
```